docs/motivation.md

```diff
@@ -0,0 +1,553 @@
+}
+// here C will inherit the default implementation of print
+```
+Here, the definition `ToString` contains a default implementation for the print
+method.
+
+* Question: Should traits be allowed to define default fields?
+
+###### External trait inheritance
+A possibility is to allow externally declared trait inheritance, i.e.
+declaration of trait inheritance can be separated from class definition.
+```verona
+// file 1
+
+// class C definition
+class C {
+    toString(s: C) : String
+    ...
+}
+
+// file 2
+// declaration that C implements ToString together with its default print method
+class C : ToString
+```
+
+
+#### Self types
+In Verona, the Self type represents the concrete type of which the object is
+part. When defining class types, Self mereley acts as an alias for the
+instantiation of this class type itself:
+```verona
+class C {
+    m1(self: Self) : Self
+    m2(self: C) : C
+    // in verona, these two method signatures are equivalent up to method name.
+}
+```
+
+For trait types, Self acts as as a reference to the underlying concrete type of the instantiation. This allows us to write e.g.
+```verona
+type Collection[T] = {
+    add(self: Self, e: T) : Self
+}
+```
+where the `add` method returns something of the concrete type, e.g.
+`list.add(x)` returns the type `List[T]` rather than `Collection[T]`.
+
+
+## Dynamic and static dispatch
+Method calls can be dynamically or statically dispatched based on the type of
+the receiver. If the receiver is a concrete class type, the compiler will
+produce a static dispatch. If on the other hand, it is something more complex, e.g., a
+disjunction type or a trait type, the method call will be dispatched
+dynamically. The compiler will of course try to optimize if the complex type is
+simple enough, e.g., `C | D` in the example below, but this happens on a case by
+case basis and the language specification will promise no such thing.
+```verona
+// classes C and D has method toString
+let x : C = C::create()
+let s = x.toString() // static dispatch, since C is a concrete (closed) type
+
+let y : C | D = if (...) { C::create() } else { D::create() }
+let t = y.toString() // dynamic dispatch, since C | D is a complex (albeit closed) type
+// it could however be optimized into something like this, since C | D is a closed world type
+let t = match y with {
+    C => y.toString() // allows method call inlining since call is now static
+    D => y.toString()
+}
+
+let z : ToString = if (...) { C::create() } else { D::create() }
+let u = z.toString() // dynamic dispatch, since ToString is an open type
+```
+
+### Type predicates
+Type predicates allow us to describe assumptions about subtyping. These can be
+written on both type and method level, and both has their distinct uses.
+Specifically, predicates on both type and method level allow us to express
+F-bounded polymorphism, while method level predicates are further useful in
+reducing code duplication, somewhat akin to type classes.
```

Comment:
Calling out "what something isn't" leads to confusion. Either the audience doesn't have that context, or worse yet, has it - and makes unfounded additional assumptions! Go/Rust/C# docs are very good at explaining without reference to other languages or out-of-language terminology.